internal/planner/planner.go

```diff
@@ -1650,11 +1656,14 @@ func (p *Planner) planRefData(virtual *ruletrie, base *baseptr, ref ast.Ref, ind
 	}
 
 	p.ltarget = base.local
+	return p.planBaseScanWithExclude(base.path, ref, index, lexclude, iter)
+}
 
-	// Perform a scan of the base documents starting from the location referred
-	// to by the data pointer. Use the set we built above to avoid revisiting
-	// sub trees.
-	return p.planRefRec(base.path, 0, func() error {
+// Perform a scan of the base documents starting from the location referred
+// to by the 'path' data pointer. Use the set (planned into 'lexclude') to
+// avoid revisiting sub trees.
+func (p *Planner) planBaseScanWithExclude(path ast.Ref, ref ast.Ref, index int, lexclude *ir.Local, iter planiter) error {
```

Comment:
💭 I had factored this out because I thought I'd end up calling it above, but then I've opted to go with the recursive call (with virtual set to nil) instead. I guess it still doesn't hurt, but I'd happily drop the commit if anyone leans the other way 😄

Reply:
I'm fine with breaking it into a separate function but perhaps we should name it so that it's more obvious? `planRefDataBaseScan` or something?